tests/ingress/ingress_test.go

```diff
@@ -0,0 +1,189 @@
+	IngressValues     = `baseDomain: %s
+`
+	HelloWorldValues = `replicaCount: 1
+ingress:
+  enabled: true
+  annotations:
+    kubernetes.io/ingress.class: nginx
+  paths:
+    - "/"
+  hosts:
+    - %s
+`
+)
+
+func Test_Ingress(t *testing.T) {
+	var err error
+
+	ctx := context.Background()
+
+	cpCtrlClient, err := ctrlclient.CreateCPCtrlClient(ctx)
+	if err != nil {
+		t.Fatalf("error creating CP k8s client: %v", err)
+	}
+
+	logger, err := micrologger.New(micrologger.Config{})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	clusterID, exists := os.LookupEnv("CLUSTER_ID")
+	if !exists {
+		t.Fatal("missing CLUSTER_ID environment variable")
+	}
+
+	cpKubeConfig, err := ctrlclient.GetCPKubeConfig(ctx)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	tcKubeConfig, err := ctrlclient.GetTCKubeConfig(ctx)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	clusterList := &capiv1alpha3.ClusterList{}
+	err = cpCtrlClient.List(ctx, clusterList, client.MatchingLabels{capiv1alpha3.ClusterLabelName: clusterID})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	baseDomain := strings.TrimPrefix(clusterList.Items[0].Spec.ControlPlaneEndpoint.Host, "api.")
+	appEndpoint := fmt.Sprintf("%s.%s", helloWorldAppName, baseDomain)
+
+	var appTest apptest.Interface
+	{
+		runtimeScheme := runtime.NewScheme()
+		appSchemeBuilder := runtime.SchemeBuilder{
+			capiv1alpha3.AddToScheme,
+			capzv1alpha3.AddToScheme,
+			expcapiv1alpha3.AddToScheme,
+			expcapzv1alpha3.AddToScheme,
+			corev1.AddToScheme,
+		}
+		err := appSchemeBuilder.AddToScheme(runtimeScheme)
+		if err != nil {
+			t.Fatal(err)
+		}
+		appTest, err = apptest.New(apptest.Config{
+			KubeConfig: string(cpKubeConfig),
+			Logger:     logger,
+			Scheme:     runtimeScheme,
+		})
+		if err != nil {
+			t.Fatal(err)
+		}
+	}
+
+	apps := []apptest.App{
+		{
+			AppCRNamespace:     clusterID,
+			AppOperatorVersion: "1.0.0",
+			CatalogName:        "giantswarm",
+			KubeConfig:         string(tcKubeConfig),
+			Name:               "nginx-ingress-controller-app",
+			Namespace:          "kube-system",
+			ValuesYAML:         fmt.Sprintf(IngressValues, baseDomain),
+			Version:            "1.11.0",
+			WaitForDeploy:      true,
+		},
+		{
+			AppCRNamespace:     clusterID,
+			AppOperatorVersion: "1.0.0",
+			CatalogName:        "default",
+			KubeConfig:         string(tcKubeConfig),
+			Name:               helloWorldAppName,
+			Namespace:          "default",
+			ValuesYAML:         fmt.Sprintf(HelloWorldValues, appEndpoint),
+			Version:            "0.3.0",
+			WaitForDeploy:      true,
+		},
+	}
+	err = appTest.InstallApps(ctx, apps)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	pod, err := createPodThatSendsHttpRequestToEndpoint(ctx, cpCtrlClient, clusterID, appEndpoint)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	o := func() error {
+		objectKey, err := client.ObjectKeyFromObject(pod)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+		scheduledPod := &corev1.Pod{}
+		err = cpCtrlClient.Get(ctx, objectKey, scheduledPod)
```

Comment:
That's pretty clever :clap: :rocket:

Reply:
![](https://media1.tenor.com/images/8f0f2655a96629f51ce6ba658e2c4723/tenor.gif?itemid=12687265)